### PR TITLE
[ESBJAVA-5044] Error while connecting to samba server - A duplicate name exists on the network

### PR DESCRIPTION
**Pull request by milindaperera**
The hostname resolution for SMB file is failed due to a limitation in the jcifs library. hence need to resolve the hostname of SMB URI in Synapse level. Hostname resolution is done at deployment to improve performance and if the user needs to dynamically resolve the hostname (in case required to adapt with /etc/hosts changes), we will introduce new property "**transport.vfs.ResolveHostsDynamically**"

Property "transport.vfs.ResolveHostsDynamically" (Boolean):
- **true**: resolve hostname at runtime (Only effective for SMB protocol)
- **false** (Default): resolve hostname at deployment

Fix Jira URL: https://wso2.org/jira/browse/ESBJAVA-5044
Note: Merge https://github.com/wso2/product-ei/pull/769 befor merging this

### CODE REVIEW

File: modules/commons/src/main/java/org/apache/synapse/commons/vfs/VFSUtils.java [hunk 1]
```diff
@@ -483,6 +483,121 @@ public static FileSystemOptions attachFileSystemOptions(Map<String, String> opti
         return opts;
     }
 
+    /**
+     * Function to resolve hostname of the vfs uri
+     * @param uri
+     * @return
+     * @throws FileSystemException
+     * @throws UnknownHostException
+     */
```
Comment (abeykoon) [inline modules/commons/src/main/java/org/apache/synapse/commons/vfs/VFSUtils.java]: Shall we complete the comment?

File: modules/commons/src/main/java/org/apache/synapse/commons/vfs/VFSUtils.java [hunk 2]
```diff
@@ -483,6 +483,121 @@ public static FileSystemOptions attachFileSystemOptions(Map<String, String> opti
...
+     * Extracts the hostname from a URI.  The scheme://userinfo@ part has
+     * been removed.
+     * extracted hostname will be reoved from the StringBuilder
+     */
+    private static String extractHostName(StringBuilder name) {
+        final int maxlen = name.length();
+        int pos = 0;
+        for (; pos < maxlen; pos++) {
+            char ch = name.charAt(pos);
+            if (ch == '/' || ch == ';' || ch == '?' || ch == ':'
```
Comment (abeykoon) [inline modules/commons/src/main/java/org/apache/synapse/commons/vfs/VFSUtils.java]: Should we describe what we do here?

File: modules/transports/core/vfs/src/main/java/org/apache/synapse/transport/vfs/PollTableEntry.java [hunk 3]
```diff
@@ -197,43 +236,59 @@ public Integer getFileProcessingCount() {
...
+            this.moveAfterProcess = resolveHostAtDeployment(moveAfterProcess);
         }
     }
 
     public String getMoveAfterErrors() {
+        if (resolveHostsDynamically) {
+            try {
+                return VFSUtils.resolveUriHost(moveAfterErrors);
+            } catch (UnknownHostException | FileSystemException e) {
+                log.warn("Unable to resolving hostname of transport.vfs.MoveAfterErrors: " +
```
Comment (abeykoon) [inline modules/transports/core/vfs/src/main/java/org/apache/synapse/transport/vfs/PollTableEntry.java]: Is this a Warning or an actual error?

File: modules/transports/core/vfs/src/main/java/org/apache/synapse/transport/vfs/PollTableEntry.java [hunk 4]
```diff
@@ -138,10 +146,25 @@ public PollTableEntry(boolean fileLocking) {
     }
 
     public String getFileURI() {
-        return fileURI;
+        if (resolveHostsDynamically) {
+            try {
+                return VFSUtils.resolveUriHost(fileURI);
+            } catch (UnknownHostException | FileSystemException e) {
+                log.warn("Unable to resolving hostname of transport.vfs.FileURI : " + VFSUtils.maskURLPassword(fileURI), e);
```
Comment (vinok88) [inline modules/transports/core/vfs/src/main/java/org/apache/synapse/transport/vfs/PollTableEntry.java]: Shall we change this to "Unable to resolve the hostname of the ...."

File: modules/commons/src/main/java/org/apache/synapse/commons/vfs/VFSUtils.java (hunk 1 above)
Comment (milindaperera) [inline modules/commons/src/main/java/org/apache/synapse/commons/vfs/VFSUtils.java]: Thanks, will update

File: modules/commons/src/main/java/org/apache/synapse/commons/vfs/VFSUtils.java (hunk 2 above)
Comment (milindaperera) [inline modules/commons/src/main/java/org/apache/synapse/commons/vfs/VFSUtils.java]: Thanks for pointing, I stole the logic from wso2-commons-vfs :-D 
https://github.com/wso2/wso2-commons-vfs/blob/master/core/src/main/java/org/apache/commons/vfs2/provider/HostFileNameParser.java#L194

I'll describe the logic in comment as you suggest

File: modules/transports/core/vfs/src/main/java/org/apache/synapse/transport/vfs/PollTableEntry.java (hunk 4 above)
Comment (milindaperera) [inline modules/transports/core/vfs/src/main/java/org/apache/synapse/transport/vfs/PollTableEntry.java]: Thanks, will update the message

File: modules/transports/core/vfs/src/main/java/org/apache/synapse/transport/vfs/PollTableEntry.java (hunk 3 above)
Comment (milindaperera) [inline modules/transports/core/vfs/src/main/java/org/apache/synapse/transport/vfs/PollTableEntry.java]: According to pattern used in this module, warning is used in most similar cases. Here warn is used due to this exception doesn't halt the continiuation of the functionality and just to find root causse if failed later.

**Pull request merged**